rust/src/smb/detect.rs

```diff
@@ -201,3 +206,433 @@ pub extern "C" fn rs_smb_tx_get_dce_iface(state: &mut SMBState,
     }
     return 0;
 }
+
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_smb_cmd_match(
+    tx: &mut SMBTransaction, cmd_data: &mut SmbCmdData,
+) -> u8 {
+
+    let version = tx.vercmd.get_version();
+    let cmd;
+    if version == 1 {
+        cmd = tx.vercmd.get_smb1_cmd().1.into();
+    } else {
+        cmd = tx.vercmd.get_smb2_cmd().1;
+    }
+
+    SCLogDebug!("rs_smb_cmd_match: version {} cmd {}", version, cmd);
+
+    if let Some(set) = cmd_data.0.get(&version) {
+        if set.contains(&cmd) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_smb_cmd_parse(carg: *const c_char) -> *mut c_void {
+    if carg.is_null() {
+        return std::ptr::null_mut();
+    }
+    let arg = match CStr::from_ptr(carg).to_str() {
+        Ok(arg) => arg,
+        _ => {
+            return std::ptr::null_mut();
+        }
+    };
+
+    match parse_cmd_data(arg) {
+        Ok(detect) => Box::into_raw(Box::new(detect)) as *mut _,
+        Err(_) => std::ptr::null_mut(),
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_smb_cmd_free(ptr: *mut c_void) {
+    if ptr != std::ptr::null_mut() {
+        std::mem::drop(Box::from_raw(ptr as *mut SmbCmdData));
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct SmbCmdData (HashMap<u8, HashSet<u16>>);
```

Comment:
Could you please comment what is this structure ? With regards to SMB1 and SMB2 ?